File: algo_problems/q141-160/q152.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0

        r = nums[0]
        cmax = r
        cmin = r
        for num in nums[1:]:
            if num < 0:
                cmax, cmin = cmin, cmax
            cmax = max(num, cmax * num)
            cmin = min(num, cmin * num)
            r = max(r, cmax)
        return r


if __name__ == '__main__':
    print(Solution().maxProduct([1, 2, 3]))
    print(Solution().maxProduct([1, 2, -3]))
    print(Solution().maxProduct([1, 0, -3, -1]))
    print(Solution().maxProduct([-2]))

File: algo_problems/q141-160/q152.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
        	return 0;
        }

        int r = nums[0];
        int imax, imin;
        imax = imin = r;
        for (int i = 1; i < nums.length; i++) {
        	if (nums[i] < 0) {
        		int tmp = imax;
        		imax = imin;
        		imin = tmp;
        	}
        	imax = Math.max(nums[i], nums[i]*imax);
        	imin = Math.min(nums[i], nums[i]*imin);
        	r = Math.max(r, imax);
        }
        return r;
    }
}